### ISSUE TITLE
[a11y] Issues with the Modal component

### ISSUE THREAD

**Issue by dpersing** [Shopify/polaris-react#795]
## Issue summary

The [Modal](https://polaris.shopify.com/components/navigation/tabs#navigation) component examples have a few accessibility issues that may impact users who rely on Windows High Contrast and/or screen readers, as well as keyboard users:

### Focus

* ~The modal does not receive focus automatically in Chrome. Chrome users must tab once to access the modal content.~ (Can no longer replicate issue in Chrome)
* It's possible to tab out of the modal before it is closed. (PR here https://github.com/Shopify/polaris-react/pull/2397)
* ~When the modal is closed with the keyboard, either by activating the `X` button or via the <kbd>esc</kbd> key, where focus moves is not immediately apparent to the user. Pressing <kbd>tab</kbd> after the modal closes causes focus to appear on the control that launched the button, but users will generally expect focus to return to the control that launched the interaction when it closes. (This looks like it should be working in code?)~ (Fixed in https://github.com/Shopify/polaris-react/pull/2140)

### Semantic structure

* ~The button that closes the modal has no text equivalent for the icon. Screen reader users will not be able to tell what the button does.~ (Fixed in #1582.)
* ~Also (while not an issue with the component itself), the text input in the "Modal with primary action" example has no label. The `<input>` has an `aria-labelledby` attribute, but its value doesn't match an `id` on the page.~ (Fixed in https://github.com/Shopify/polaris-react/pull/2401)
* ~If no `title` is provided, the `aria-labelledby` attribute of the modal points to an invalid `id`.~ (Fixed in PR https://github.com/Shopify/polaris-react/pull/2115.)

The focus issue may also relate to a weird deal with NVDA/Firefox, where the virtual buffer is still able to read the modal content even after it's removed from the DOM. (I tested with NVDA/Chrome as well, just to see what it would do, and it reads as expected.) Examples I've tested where focus goes directly to an onscreen element don't have this issue, since that updates the buffer.

### High Contrast support

* ~The shape and size of the modal is conveyed through a background color and `box-shadow`, which is not visible when the component is viewed in High Contrast. Users may not be exactly sure of the shape and size of the modal, and what it overlays.~ (Fixed in PR https://github.com/Shopify/polaris-react/pull/2114.)

![Basic modal example](https://user-images.githubusercontent.com/1462085/41789834-7e880cdc-7605-11e8-9d75-c6e278944f61.png)

![Modal viewed in High Contrast in Windows 10](https://user-images.githubusercontent.com/1462085/41789779-5234c878-7605-11e8-9a39-4d5ca2723d2a.png)

#### `X` button (HTML)

```
<button class="Polaris-Modal-Header__CloseButton">
  <span class="Polaris-Icon Polaris-Icon--colorInkLighter Polaris-Icon--isColored">
    <svg class="Polaris-Icon__Svg" viewBox="0 0 20 20" focusable="false" aria-hidden="true">
    <!-- ... --></svg>
  </span>
</button>
```

#### Modal container styles (CSS)

```
.Polaris-Modal-Dialog__Modal {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: -ms-flexbox;
  display: flex;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100vw;
  height: 100%;
  max-height: calc(100vh - 60px);
  background: #fff;
  box-shadow: 0 31px 41px 0 rgba(32,42,53,.2), 0 2px 16px 0 rgba(32,42,54,.08);
}
```

#### Input example (HTML)

```
<input id="TextField4" class="Polaris-TextField__Input" aria-labelledby="TextField4Label" aria-invalid="false" value="https://polaris.shopify.com/">
```

### Button and label text

1. Navigate to the [Modal](https://polaris.shopify.com/components/navigation/tabs#navigation) component examples.
1. Using VoiceOver or NVDA, open each modal and read through the modal content to hear how it is identified.
1. Use the <kbd>tab</kbd> key to move through controls in the modals to hear how they are read.

### High Contrast

1. Navigate to the [Modal](https://polaris.shopify.com/components/navigation/tabs#navigation) component examples.
1. Open the default modal, and do a visual inspection.
1. Turn on any [High Contrast theme in Windows settings](https://support.microsoft.com/en-us/help/13862/windows-use-high-contrast-mode) and redo the visual inspection.

### Keyboard

1. Navigate to the [Modal](https://polaris.shopify.com/components/navigation/tabs#navigation) component examples.
1. <kbd>tab</kbd> to the `Open` button and press <kbd>return/enter</kbd> or <kbd>space</kbd> to activate it.
1. To close the modal, <kbd>tab</kbd> to the `X` button and activate it.
1. Repeat, closing the modal instead with the <kbd>esc</kbd> key.

### Expected

- Buttons will have text that describes their purpose.
- Features will be easy to discern without relying on color.
- Focus will be managed to and from modals since they disrupt the natural DOM/keyboard focus flow.
- `<input>` elements will have a valid label.

### Actual

- Focus does not always go to the modal.
- The `X` button icon has no text equivalent.
- The modal container dimensions are difficult to see without its background styles when viewed in Windows High Contrast.
- Focus is not always managed back to the control that activates the modal when it is closed.
- The example `<input>` has no label.

## Recommendations

- Give the `X` button an `aria-label` or visually hidden text that conveys its purpose (`aria-label="Close modal"`).
- If no `title` is provided, do not include an `aria-labelledby` attribute on the modal.
- Give the modal container a `transparent` border so when the border is made a solid color automatically in High Contrast, the edges of the modal are easy to see. If preferred, this can be applied using the `-ms-high-contrast` media query, but it shouldn't hurt to just include it normally.
- Focus management for the modal can probably be optimized. When the modal is closed, send focus directly back to the control that opens it to avoid any confusion about where focus is (and probably fix the NVDA bug too.) Focus should be managed when a user closes the modal using the "X" control as well as the Esc key.
- Give the text input example a label. A visually hidden one is totally fine since the gist of the input text is described in the preceding content.

[See the ARIA authoring practices modal example.](https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html)

There are a number of enhancements currently available in the 1.1 spec for modals, but support is either not present or buggy at present. We can [keep an eye on that progress](https://developer.paciellogroup.com/blog/2018/06/the-current-state-of-modal-dialog-accessibility/) and add those in time.

#### `X` button recommendation example (HTML)

```
<button class="Polaris-Modal-Header__CloseButton" aria-label="Close modal">
  <span class="Polaris-Icon Polaris-Icon--colorInkLighter Polaris-Icon--isColored">
    <svg class="Polaris-Icon__Svg" viewBox="0 0 20 20" focusable="false" aria-hidden="true">
    <!-- ... --></svg>
  </span>
</button>
```

### Modal border recommendation example (CSS)

```
.Polaris-Modal-Dialog__Modal {
  ...
  border: 1px solid transparent;
}
```

#### Input label recommendation example (HTML)

```
<label for="TextField4"><span class="Polaris-VisuallyHidden">Shareable link</span></label>
<input id="TextField4" class="Polaris-TextField__Input" aria-invalid="false" value="https://polaris.shopify.com/">
```

## Specifications

- **Component:** Modal
- **WCAG SCs:** [1.4.1 Use of Color (Level A)](https://www.w3.org/TR/WCAG21/#use-of-color), [2.4.3 Focus Order (Level A)](https://www.w3.org/TR/WCAG21/#focus-order), [2.4.7 Focus Visible (Level AA)](https://www.w3.org/TR/WCAG21/#focus-visible), [3.3.2 Labels or Instructions (Level A)](https://www.w3.org/TR/WCAG21/#labels-or-instructions), [4.1.2 Name, Role, Value (Level A)](https://www.w3.org/TR/WCAG21/#name-role-value)
- **Testing done with:**
  - Keyboard
  - VoiceOver/Safari/macOS
  - NVDA/Firefox/Windows 10
  - Windows High Contrast/Edge/Windows 10

**Comment by dpersing** [comment]
We can keep an eye on [`inert` support](https://www.chromestatus.com/feature/5703266176335872) in regard to modals as well.

**Comment by dpersing** [comment]
>The button that closes the modal has no text equivalent for the icon. Screen reader users will not be able to tell what the button does.

This item will be addressed by PR #1582.

**Comment by chloerice** [comment]
> Focus

> The modal does not receive focus automatically in Chrome. Chrome users must tab once to access the modal content.

> It's possible to tab out of the modal before it is closed.

> When the modal is closed with the keyboard, either by activating the X button or via the esc key, where focus moves is not immediately apparent to the user. Pressing tab after the modal closes causes focus to appear on the control that launched the button, but users will generally expect focus to return to the control that launched the interaction when it closes. (This looks like it should be working in code?)

When the first audit was done, our overlay components like Modal and Popover rendered only their activators and required interaction to be visible. Are the problems outlined around focus in the Modal examples still a concern now that our overlays default to being `open` or `active` @dpersing ?

**Comment by dpersing** [comment]
> Are the problems outlined around focus in the Modal examples still a concern now that our overlays default to being open or active

I believe so. I think folks should be able to close the examples and reopen them to see the expected behavior.

**Comment by dpersing** [comment]
Some accidental testing within the Vault and admin shows that if the Modal component appears in a page using the Nav component, focus goes to the "Skip to content" link as part of the modal focus order, which is unexpected. My initial thought that @LauraAubin's work to keep focus from going to the browser chrome should help this. 🤔

**Comment by LauraAubin** [comment]
Closing this since the last task was completed with https://github.com/Shopify/polaris-react/pull/2555
